### PR TITLE
add recipe amread-mode

### PR DESCRIPTION
**Pull request by stardiviner**
### Brief summary of what the package does

A minor mode helping user reading like Spritz.

### Direct link to the package repository

https://github.com/stardiviner/amread-mode

### Your association with the package

maintainer

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [ ] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I have confirmed some of these without doing them

### CODE REVIEW

**Comment by terlar** [comment]
How does this compare to spray? https://melpa.org/#/spray

**Comment by stardiviner** [comment]
@terlar Similar with spray, but still different, It is very much like Spritz which only display one word in buffer. But I just display property on word and move forward. I keep buffer content what it is.

**Comment by stardiviner** [comment]
I try to record a screencast for demo, but I found I can't record it correctly.

**Comment by terlar** [comment]
Thank you, that would definitely help, I will also try it when I have time. I recently found the tool [peek](https://github.com/phw/peek) worked well to record such videos. Have you tried it?

**Comment by stardiviner** [comment]
I tried `peek`, still recording weird. Can't record the property moving animation. Weird. @terlar Can you record one to help me if you can? Thanks for this.

**Comment by terlar** [comment]
Here you go:
Peek 2020-01-30 23-18

Is this looking like it should for you?

I have some issue with using the keys, it worked a few times, but then I just got buffer is read-only on every key press q/SPC. So I had to quit by doing `M-x`.

**Comment by stardiviner** [comment]
@terlar Thanks for your screencast gif and testing out. Can you tell me your latest using commit version? I bind `q` to quit amread-mode. and bind `[SPC]` to pause-or-resume. In the design, amread-mode will set buffer to read-only-mode, and exit read-only-mode when exit amread-mode.

**Comment by stardiviner** [comment]
I updated code, added commands to enable/disable amread-mode, then bind `q` to `amread-mode-disable` explicitly. Should fix your problem now.

File: recipes/amread-mode [hunk 1]
```diff
@@ -0,0 +1 @@
+(amread-mode :repo "stardiviner/amread-mode" :fetcher github :files ("amread-mode.el"))
```
Comment (riscy) [inline recipes/amread-mode]:
```suggestion
(amread-mode :repo "stardiviner/amread-mode" :fetcher github)
```

Comment (stardiviner) [reply recipes/amread-mode]: Thanks

**Comment by stardiviner** [comment]
> amread-mode.el:44:1:Warning: Unused lexical variable `length'

I fixed this issue by remove this unused variable.

> prefer https over http

fixed too

Thanks for your tips. And other LICENSE  text, I'm not very clear understanding.

**Comment by riscy** [comment]
> And other LICENSE text, I'm not very clear understanding.

No need to act on any of that stuff. :)

**Comment by alphapapa** [comment]
FYI, for recording screencasts: https://github.com/alphapapa/emacs-package-dev-handbook#animations--screencasts

**Comment by conao3** [comment]
Hi!
I'm not a MELPA maintainer, but here's some feedback for this package.
I hope below those comments help you and your package.

1. Please consider adding Cask file is good to easy resolve dependency, build, packaging etc.
2. Please add .gitignore file is good to ignore git stage *.elc files and .cask folder.
3. Please remove unneeded `prog1` because return value is not used.
4. Please consider removing `amread-mode-{enable, disable}` function.
  Minor mode is safely enabling and disabling minor mode, `amread-mode` is already interactive toggle function.

**Comment by stardiviner** [comment]
@conao3 Thanks, I added Cask file now. And .gitignore, and followed you advice by some modifications.

**Comment by riscy** [comment]
A couple last items, but I think we're almost there:

- You use the keyword 'tools' but per `finder-known-keywords` that keyword is for "programming tools".  (Many of these keywords feel like they come from a different Emacs era, but I think we can still conform.)  I feel like "convenience" is a good match, or possibly "wp" (word processing).
- I don't see "speed-reading" as a keyword anywhere -- but isn't that sort of what this is?  I would have trouble finding this if I were searching for such facilities.
- Your `:lighter` shouldn't have a trailing space -- only a leading space -- and probably shouldn't have a space in it.  What about just `" amread"`, which makes it really clear which minor mode is active?
- Should `amread-start` and `amread-stop` get an `;;;###autoload` cookie?
- `(define-key map [remap keyaobrd-quit] 'amread-mode-disable)` -> `keyboard-quit`
- If narrowing is in effect on a buffer, I wonder if you want to use `(goto-char (point-min))` instead of `(goto-char 0)` (on line 71)

**Comment by stardiviner** [comment]
@riscy Hi, I updated amread-mode, actually I updated many days ago on my laptop, but I'm offline, so now pushed up. Thanks for point out.

**Comment by riscy** [comment]
Awesome, thank you. :)

I'm seeing a couple more tiny little nags:

### amread-mode.el ###

byte-compile-file (using Emacs 26.3):
```
amread-mode.el:44:1:Warning: Unused lexical variable `length'
```

checkdoc (using version 0.6.1):
- No issues!

package-lint-current-buffer (using version 20200226.2001):
```
1 issue found:
1:0: warning: The package summary should not end with a period.
```

And while I'm having a peek, you could sharp-quote the keybinding functions (which will help the byte-compiler detect if they get misnamed in the future), as in:
```elisp
    (define-key map (kbd "q") #'amread-mode-quit)
    (define-key map (kbd "SPC") #'amread-pause-or-resume)
    (define-key map [remap keyboard-quit] #'amread-mode-disable)
```
...and actually, applying that change revealed to me that `#'amread-mode-disable` does not exist as a function. :)

**Comment by stardiviner** [comment]
> amread-mode.el:44:1:Warning: Unused lexical variable `length'

This is necessary step to move point forward. If I remove it, the point will not move forward.

Other modifications done.

**Comment by riscy** [comment]
> This is necessary step to move point forward

I see.  You have a couple options here -- `let` forms let you use `progn` so you can:
```elisp
  (let* ((begin (point))
         ;; move point forward. NOTE This forwarding must be here before moving overlay forward.
         (end (progn
                (skip-chars-forward "^\s\t\n—")
                (skip-chars-forward "—")
                (point)))
```
or, if for some reason you want to keep that length calculation intact, you can tell the compiler to ignore it by prefixing with `_`:
```elisp
  (let* ((begin (point))
         ;; move point forward. NOTE This forwarding must be here before moving overlay forward.
         (_length (+ (skip-chars-forward "^\s\t\n—") (skip-chars-forward "—")))
         (end (point)))
```

I notice you still haven't fixed the issue where `amread-mode-disable` isn't defined -- please re-read that carefully.

**Comment by riscy** [comment]
> please re-read that carefully.

Actually it's me who needs to re-read carefully :)  I had an old amread buffer open.

**Comment by stardiviner** [comment]
:) I indeed renamed deleted command name.
I select the second solution incause I need to use the calculated length in future.

**Comment by stardiviner** [comment]
Any update on this? I think this is stable for MELPA users. :)

**Comment by riscy** [comment]
Sorry for the delay, I'll take a look again soon!

**Comment by riscy** [comment]
Looking good -- I noticed you recently added some cl code so we're here with the byte-compile:
byte-compile-file (using Emacs 26.3):
```
amread-mode.el:136:1:Warning: the following functions are not known to be
    defined: incf, decf
```
To resolve these, you should `(require 'cl-lib)` and use `cl-incf` and `cl-decf` (`cl` is deprecated, so we prefer the `cl-lib` equivalents).

**Comment by riscy** [comment]
(Ping me and this is a quick merge after that.)

**Comment by stardiviner** [comment]
Sorry for delay, I added `cl-lib` requirement, and updated GitHub commit. Thanks for reviewing my code :)

**Comment by riscy** [comment]
Non-blockers (sorry for all the nags), but these just popped up.  This is just an fyi:

checkdoc (using version 0.6.1):
```
Warning (emacs): 
amread-mode.el:99: Lisp symbol `amread-mode' should appear in quotes
Warning (emacs): 
amread-mode.el:104: Lisp symbol `amread-mode' should appear in quotes
Warning (emacs): 
amread-mode.el:109: Lisp symbol `amread-mode' should appear in quotes
Warning (emacs): 
amread-mode.el:121: Lisp symbol `amread-mode' should appear in quotes
```

**Comment by stardiviner** [comment]
Oh, I have not re-run `checkdoc` after added new code. Thanks for merging. I updated my docstrings. And I run `checkdoc`, now all passed. :100:

**Pull request merged**